### PR TITLE
Detect personal-space folder and move to very top, decode url encoded …

### LINKED ISSUE #29: Add "personal space" for teachers
Teachers should have a folder called "personal space" or something similar at the very top of their root-level folders. It might also look a little bit different color-wise.
Additionally there should be a file upload button inside this folder.

### PR DESCRIPTION
**Pull request by theopfr**
Closes #29 
In order to display the personal folder for tutors we need to fetch the root folders as usual and the when we detect a folder called "Persönliche Ablage" it should be moved to the very top of the folders list so it's shown at the first folder. To test this please create a folder called ``Persönliche Ablage`` in NextCloud.

Now I also decode the file and folder names. Previously they were url-encoded. for example all spaces where "%20", eg. ``Nextcloud%20Logo.jpg``, now it will be displayed as ``Nextcloud 20Logo.jpg``, this also makes Umlaute (äöü) etc work.

### Edit:
I forgot to add the functionality that users can also upload files into their personal space. So now there is an upload button with which you can upload files in there. The question remains if I also should add a "create folder" button, in order to create a sub folder. Let me now. We can also do this in the future if we need to.

### CODE REVIEW

File: app/components/FileList.tsx [hunk 1]
```diff
@@ -14,9 +14,25 @@ const FileList = (props: FileListType) => {
     const [files, setFiles] = useState<FileCardType[]>([]);
     const [folders, setFolders] = useState<FolderCardType[]>([]);
 
+    const PERSONAL_SPACE_FOLDER_NAME = "Pesönliche Ablage";
```
Comment (cma5) [inline app/components/FileList.tsx]: I think, it should be "Persönliche Ablage" (with an R) ;)

File: app/components/FileList.tsx [hunk 2]
```diff
@@ -14,9 +14,25 @@ const FileList = (props: FileListType) => {
...
+            folders.unshift(personalSpaceFolder);
+        }
+
+        return folders;
+    }
+
     useEffect(() => {
         setFiles(props.files);
-        setFolders(props.folders);
+        const rearangedFolders = personalFolderToTop(props.folders);
```
Comment (cma5) [inline app/components/FileList.tsx]: It's called "rearranged" with 2 R

**Review by cma5 — CHANGES_REQUESTED** [review]

File: app/components/FileList.tsx (hunk 2 above)
Comment (theopfr) [reply app/components/FileList.tsx]: Thanks, changed it

File: app/components/FileList.tsx (hunk 1 above)
Comment (theopfr) [reply app/components/FileList.tsx]: Indeed! Changed it

**Comment by theopfr** [comment]
I forgot to add the functionality that users can also upload files into their personal space. So now there is an upload button with which you can upload files in there. The question remains if I also should add a "create folder" button, in order to create a sub folder. Let me now. We can also do this in the future if we need to. @cma5 

Please review again and test the file upload. It was also once again a pain to implement..

File: app/components/FileList.tsx [hunk 3]
```diff
@@ -14,9 +14,23 @@ const FileList = (props: FileListType) => {
     const [files, setFiles] = useState<FileCardType[]>([]);
     const [folders, setFolders] = useState<FolderCardType[]>([]);
 
+    // Moves the "personal space" folder to the very top if it exists
+    const personalFolderToTop = (folders: FolderCardType[]) => {
+        const index = folders.findIndex(folder => decodeURIComponent(folder.folderName) === PERSONAL_SPACE_FOLDER_NAME);
```
Comment (cma5) [inline app/components/FileList.tsx]: For ease of use, we should create the folder automatically on startup if it doesn't exist

**Review by cma5 — APPROVED** [review]
All in all, it works pretty good. I'm thinking if we could safe ui work if we use premade components form react native elements. This could make our lives easier, and they probably have stuff like add folder functions etc as well

File: app/components/FileList.tsx (hunk 3 above)
Comment (theopfr) [reply app/components/FileList.tsx]: definetly, I just don't know at what stage this initial folder structur should be created. In a normal application the backend would do tha of course but we don't have one and as far as we know initializing a folder structure on the container start up doesnt work, right?

Then we should maybe create a function that runs once when clicking on the ``Dateien`` tab and which initializes the folder structure if t wasn't done before. Of course it's very weird to do that in the frontend but for our rest-less prototype it's a good solution.

Comment (theopfr) [reply app/components/FileList.tsx]: Which components are you thinking off? Like file/folder cards?

File: app/components/FolderContentScreen.tsx [hunk 4]
```diff
@@ -77,7 +82,32 @@ const FolderContentScreen = (props: NativeStackScreenProps<RootStackParamList, "
...
         }
         catch {
-            Alert.alert("File upload failed!");
+            Alert.alert("Failed to upload file!");
+        }
+    }
+
+    const addFolderHandler = async () => {
+        try {
+            await createFolder(`${props.route.params?.folderURL}/${inputText}`).then((status) => {
```
Comment (cma5) [inline app/components/FolderContentScreen.tsx]: We should delete special characters when creating a folder. Didn't test everything but for example if you put in a [ the server doesn't create a folder and the user also doesn't get feedback that no folder was created

Comment (theopfr) [reply app/components/FolderContentScreen.tsx]: Actually for me nextcloud handles folders named `[` well. But yes we don't want such filenames allowed therefore users can now only create folder names with lating letters, numbers, "-", "_" and spaces. When they enter something else they will see a error message explaining whats allowed.

File: app/components/FolderContentScreen.tsx [hunk 5]
```diff
@@ -37,23 +54,165 @@ const FolderContentScreen = (props: NativeStackScreenProps<RootStackParamList, "
...
+                            <Feather name="x" size={20} color="black" />
+                        </TouchableOpacity>
+
+                        <Text style={styles.modalTitle}>Ordner Name</Text>
+                        <TextInput
+                            style={styles.textInput}
+                            placeholder="Wähle einen Ordner Namen..."
+                            value={newFolderName}
+                            onChangeText={(text: string) => {
+                                if (/[^a-zA-Z0-9-_ ]/.test(text)) {
```
Comment (cma5) [inline app/components/FolderContentScreen.tsx]: Hmm, it's not so easy... I would also like to have Ö ö Ä ä Ü ü and ß in my foldername.

**Review by cma5 — CHANGES_REQUESTED** [review]

File: app/components/FolderContentScreen.tsx (hunk 5 above)
Comment (theopfr) [reply app/components/FolderContentScreen.tsx]: ahh I simply forgot, one second

Comment (theopfr) [reply app/components/FolderContentScreen.tsx]: now all kind of letters are allowed, I also set a max folder name length (128).

**Review by cma5 — APPROVED** [review]
Thanks for the changes. Works flawlessly!

**Pull request merged**